### PR TITLE
Task/dss121 aks improvements

### PR DESCRIPTION
**Pull request by FChataigner**
[sc-136682]
[sc-135811]
[sc-136680]

also added a "add node pool" macro to make testing the "resize node pool" one easier

### CODE REVIEW

**Review by marchelleboid — CHANGES_REQUESTED** [review]
On the resize macro, it seems you can't delete the last node pool or the "system" node pools. Should we try to catch those errors before they happen? Or let the error messages from Azure display on the UI? They seemed to be pretty clear, except for one message that says simply "Operation no supported" or something like that.

For the availability zone one, I tested on `ukwest`, and the error message is clear. Switching to the standard load balancer seems OK to me. We need to remember to update the doc then.

Subscription change seems good to me. Tested by inputing a managed identity and the current subscription (do we have access to another one?), and it worked.

File: python-lib/dku_azure/utils.py [hunk 1]
```diff
@@ -11,7 +12,16 @@ def run_and_process_cloud_error(fn):
     try:
         return fn()
     except Exception as e:
-        raise e
+        try:
+            str_e = str(e)
+        except:
+            logging.warn("Can't inspect error")
+            str_e = ''
+        if 'Availability zone is not supported in region' in str_e:
```
[the remark this answers is not in the document]
Comment (FChataigner) [reply python-lib/dku_azure/utils.py:20]: added this other error formulation

File: python-runnables/resize-node-pool/runnable.py [hunk 2]
```diff
@@ -31,22 +31,28 @@ def run(self, progress_callback):
         
         node_pool_id = self.config.get('nodePoolId', None)
         node_pool = None
-        for profile in cluster.agent_pool_profiles:
-            if profile.name == node_pool_id or (_is_none_or_blank(node_pool_id) and len(cluster.agent_pool_profiles) == 1):
+        node_pools = [node_pool for node_pool in clusters.agent_pools.list(resource_group, cluster_name)]
+        for profile in node_pools:
+            if profile.name == node_pool_id or (_is_none_or_blank(node_pool_id) and len(node_pools) == 1):
```
[the remark this answers is not in the document]
Comment (FChataigner) [reply python-runnables/resize-node-pool/runnable.py:36]: done

**Review by amandineslx — CHANGES_REQUESTED** [review]
* should we change the formulation about the default node pool name in the **Add node pool** and **Resize cluster** macros?
* when resizing a node pool, the node number in the Azure portal seems unchanged (maybe not linked to this particular PR)
* cryptic error message when triggering macro **Resize cluster** without providing a node pool name but with a number of nodes different from 0

### Tested

* build plugin archive from the branch in the PR (revision `d026430cf706354ae216629eca552a4541f984ee`)
* deploy DSS with FM on Azure in an Elastic fleet
  * start a cluster
  * install the new version of the plugin on DSS
  * trigger macro **Add node pool**
    * :warning: default advertised node pool name is **default** when the node pools get named incrementally **nodepoolX**
    * :white_check_mark: both the initial and the added one are visible in the Azure portal and the characteristics are matching what was provided
  * trigger macro **Resize cluster** without a node pool name and with a number of nodes equal to 0
    * :x: default advertised node pool name is **default**. When not providing a value, **None** is used
    * :white_check_mark: when providing the node pool name, the corresponding node pool gets deleted
  * trigger macro **Resize cluster** without a node pool name and with a number of nodes different from 0
    * :x: error raised is cryptic: `<class 'azure.core.exceptions.HttpResponseError'> : Operation returned an invalid status 'Method Not Allowed'`
    * :x: when providing the node pool name, the number of nodes in the pool is unchanged in the Azure portal
  * set the credentials identity type
    * :white_check_mark: the field to provide the Azure subscription ID displays for all options
  * open the page to create a new cluster
    * :white_check_mark: the default load balancer SKU is `Standard`
    * :warning: doc to be updated
  * provide a managed identity (DSS one) and Dataiku subscription ID to start the cluster
    * :white_check_mark: cluster is starting correctly
* deploy DSS with FM on Azure in an Elastic fleet in region `ukwest`
  * start a cluster with availability zone option activated
    * :white_check_mark: clear error is raised soon after the startup trigger
  * update plugin with archive built from the branch in the PR (revision `adea6bb4289d10f64120c132b3f245a02514fae9`)
  * add a node pool with availability zone option activated
    * :white_check_mark: clear error is raised

File: python-runnables/resize-node-pool/runnable.py [hunk 3]
```diff
@@ -31,22 +31,30 @@ def run(self, progress_callback):
...
                 node_pool = profile
         if node_pool is None:
             raise Exception("Unable to find node pool '%s'" % (node_pool_id))
 
         desired_count = self.config['numNodes']
         logging.info("Resize to %s" % desired_count)
         if desired_count == 0:
-            raise Exception("Can't delete node pool '%s'" % (node_pool_id))
+            def do_delete():
+                cluster_update_op = clusters.agent_pools.begin_delete(resource_group, cluster_name, node_pool_id)
```
Comment (marchelleboid) [inline python-runnables/resize-node-pool/runnable.py:47]: `node_pool_id` will still be `None` here if it wasn't set. You probably need to do a `if node_pool_id is None: node_pool_id = node_pool.name` after selecting the `node_pool`.

**Review by marchelleboid — CHANGES_REQUESTED** [review]

**Review by marchelleboid — APPROVED** [review]
Retested the resize and add node pool macros
* Created a GPU node pool and verified that it tried to install the driver
* Created a system node group
* Didn't exhaustively test each field in the add node pool though
* The error when trying to delete the last node pool is clear
* The error when trying to delete a system node pool is directly from the Azure client, but it's clear enough
* Tested leaving the id field blank and resizing the only node pool (deleting fails of course)

Changelog looks good to me

**Review by amandineslx — APPROVED** [review]
OK for me

### Tested

* build plugin archive from the branch in the PR (revision `070f52a5586bf1b16100bb5fa9c20855d3a7576e`)
* deploy DSS with FM on Azure in an Elastic fleet in `ukwest` region
  * start a cluster
  * install the new version of the plugin on DSS
  * trigger macro **Resize cluster** (nodePoolName='',nodesNb=0)
    * :white_check_mark: when supplying no node pool name, if only 1 node pool, the only one available is used
    * :white_check_mark: if there is only one node pool, the error message is clear
    * :white_check_mark: when providing the node pool name, the corresponding node pool gets deleted
  * trigger macro **Add node pool**
    * :white_check_mark: the additional node pool is visible in the cluster
  * trigger macro **Resize cluster** (nodePoolName='',nodesNb=0)
    * :white_check_mark: error message is clear about choosing a node pool to impact
  * trigger macro **Resize cluster** (nodePoolName='',nodesNb=2)
    * :white_check_mark: error message is clear about choosing a node pool to impact
  * trigger macro **Resize cluster** (nodePoolName='nodepool1',nodesNb=0)
    * :white_check_mark: node pool gets deleted
  * trigger macro **Resize cluster** (nodePoolName='',nodesNb=2)
    * :white_check_mark: node pool `nodepool0` get resized to 2 nodes
  * set the credentials identity type
    * :white_check_mark: the field to provide the Azure subscription ID displays for all options
  * open the page to create a new cluster
    * :white_check_mark: the default load balancer SKU is `Standard`
    * :warning: doc to be updated
  * provide a managed identity (DSS one) and Dataiku subscription ID to start the cluster
    * :white_check_mark: cluster is starting correctly
  * start a cluster with availability zone option activated
    * :white_check_mark: clear error is raised soon after the startup trigger
  * add a node pool with availability zone option activated
    * :white_check_mark: clear error is raised
 * ✅ changelog added

**Pull request merged**